cms/djangoapps/contentstore/middleware.py

```diff
@@ -0,0 +1,55 @@
+"""
+Middlewares for contentstore app.
+"""
+
+import logging
+
+from django.conf import settings
+from django.http import HttpResponse
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import CourseKey
+
+from common.djangoapps.student.auth import (
+    STUDIO_EDIT_CONTENT,
+    STUDIO_VIEW_CONTENT,
+    get_user_permissions,
+)
+from openedx.core.lib.request_utils import COURSE_REGEX
+
+logger = logging.getLogger(__name__)
+
+
+class LimitedStaffAccessMiddleware:
+    """
+    Middleware to check access for limited staff users.
+    """
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def get_course_key(self, request):
+        """
+        Extract course key from request URI if it exists.
+        """
+        match = COURSE_REGEX.match(request.build_absolute_uri())
```

Comment:
I believe this regex only works with this kind of endpoints: `.../course(s)/<course_id>` with this pattern 
```
'^(.*?/course(s)?/)(?!v[0-9]+/[^/]+)(?P<course_id>[^/+]+(/|\\+)[^/+]+(/|\\+)[^/?]+)'
```
So I still have access to other endpoints like:
` {{ CMS_DOMAIN }}/certificates/course-v1:{{ COURSE_ID }}`
` {{ CMS_DOMAIN }}/assets/course-v1:{{ COURSE_ID }}/`
and so on. I think we should use something like: 
```
'(^(.*?)(?P<course_id>course-v1:[^/+]+(\+)[^/+]+(\+)[^/?]+)'
```
or any other similar regex. I still don't understand why `COURSE_REGEX` doesn't get the course ID for all kinds of endpoints: https://github.com/openedx/edx-platform/blob/master/openedx/core/lib/request_utils.py#L18

With these changes and some other tweaks, I got to this implementation that blocked any request made by a limited staff:

```
import logging
import re
from common.djangoapps.student.roles import CourseLimitedStaffRole
from django.conf import settings
from django.http import HttpResponse
from opaque_keys import InvalidKeyError
from opaque_keys.edx.keys import CourseKey

logger = logging.getLogger(__name__)

class LimitedStaffAccessMiddleware:
    """
    Middleware to check access for limited staff users.
    """

    def __init__(self, get_response):
        self.get_response = get_response

    def get_course_key(self, request):
        """
        Extract course key from request URI if it exists.
        """
        match = re.compile(r'^(.*?)(?P<course_id>course-v1:[^/+]+(\+)[^/+]+(\+)[^/?]+)').match(request.build_absolute_uri())
        if match:
            course_id = match.group("course_id")
            try:
                return CourseKey.from_string(course_id)
            except InvalidKeyError:
                logger.exception("Invalid course key: {}".format(course_id))
        return None

    def __call__(self, request):
        user = request.user
        course_key = self.get_course_key(request)
        if settings.ROOT_URLCONF == "cms.urls" and course_key is not None:
            if CourseLimitedStaffRole(course_key).has_user(user):
                logger.exception(
                    "PermissionDenied: Failed in authenticating {}".format(user)
                )
                return HttpResponse(status=403)

        return self.get_response(request)
```

Reply:
If we're doing regex matching, shouldn't we use the same patterns that Studio's `urls.py` uses, like `COURSELIKE_KEY_PATTERN`, `LIBRARY_KEY_PATTERN`, `settings.USAGE_KEY_PATTERN`, `settings.COURSE_KEY_PATTERN`, `settings.ASSET_KEY_PATTERN`?

We'd also need to filter for usage keys and possibly asset keys, because those endpoints can implicitly encode a course without having a `course-v1:`, and those endpoints could potentially allow writes.

Reply:
I think we might be able to use more of Django's url resolver functionality to actually get the view and metadata of the match itself, so we can have more confidence that we're matching the same behavior. Looking into this...

Reply:
Yeah, I think we can do this. Here's a sample:

```python
In [1]: from django.urls import resolve
In [2]: match = resolve("/course/course-v1:Axim+PERF+2024-05-03")
In [3]: match
Out[3]: ResolverMatch(func=cms.djangoapps.contentstore.views.course.course_handler, args=(), kwargs={'course_key_string': 'course-v1:Axim+PERF+2024-05-03'}, url_name='course_handler', app_names=[], namespaces=[], route='^course/(?P<course_key_string>[^/+]+(/|\\+)[^/+]+(/|\\+)[^/?]+)?$', captured_kwargs={'course_key_string': 'course-v1:Axim+PERF+2024-05-03'})
```

Now we're not terribly consistent about whether we call things `course_id`, `course_key`, or whatever. But it doesn't really matter at this point because `match.args` and `match.kwargs` should have the values that we're calling our views with. We don't have to trust the names–each request is only going to have a few of them, so we can check if they're strings and then try to parse them as course keys, usage keys, and asset keys. See if there's anything we can use to derive a course from, and then disallow if necessary.

Reply:
@mariajgrimaldi indeed with the previous pattern we missed most of the endpoints and I also changed the limited staff validation.

Reply:
@ormsbee thanks for the new approach to key parsing it really makes sense. It looks like it covers most of the queries made to cms.

Reply:
We are planning to continue testing and I will report as soon as it is ready.